### PR TITLE
chore: refactor string match

### PR DESCRIPTION
**Pull request by zhaohuabing**
fix:  https://github.com/envoyproxy/gateway/issues/2067

* use the shared string match for stats match condition
* rename string match type in shared types, if no special requirement like 'Distinct' in Rate limit, all the other EG APIs should consider using this common string match type
* use the shared string match for rate limit - removed after discussion

### CODE REVIEW

File: api/v1alpha1/ratelimit_types.go [hunk 1]
```diff
@@ -108,49 +108,25 @@ type SourceMatch struct {
...
+
+	// Match specifies how to match against the value of the header.
 	// Do not set this field when Type="Distinct", implying matching on any/all unique
 	// values within the header.
-	//
-	// +optional
-	// +kubebuilder:validation:MaxLength=1024
-	Value *string `json:"value,omitempty"`
+	// Only one of Distinct or Match can be set.
+	Match *StringMatch `json:"match,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/ratelimit_types.go:127]: @arkodg @zirain @Xunzhuo  I plan to refactor rate limit HeaderMatch to use the shared string match. If you agree this is the right way, I'll go ahead.

Comment (arkodg) [reply api/v1alpha1/ratelimit_types.go:127]: we should keep header matches aligned to upstream as much as possible.
for this case we defined in this api because we introduced a new match type called `distinct` that doesnt exist upstream
https://github.com/kubernetes-sigs/gateway-api/blob/5fd8bec3bedc7c325bf2227da48d041a735b9b98/apis/v1beta1/httproute_types.go#L98
https://github.com/kubernetes-sigs/gateway-api/blob/5fd8bec3bedc7c325bf2227da48d041a735b9b98/apis/v1/httproute_types.go#L439

Also although sprinkling different match types affects project mantainence, it benefits end users because for each different feature we are adding a different set of kubebuilder annotations for the fields, so we providing proactive feedback on the type of matches that we support.

tldr, if the match types are identical, it makes sense to consolidate, else imo we should wait until they reach parity

Comment (zhaohuabing) [reply api/v1alpha1/ratelimit_types.go:127]: IMO, the Distinct Match type only matters when it comes to the Rate Limit. So, perhaps we could consider renaming HeaderMatch to RateLimitHeaderMatch. Let's leave it as is for now and see how we want to deal with the string match when we have new APIs in the mix.

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 2]
```diff
@@ -60,18 +60,18 @@ type ProxyPrometheusProvider struct {
...
 	//
 	// +kubebuilder:validation:Enum=RegularExpression;Prefix;Suffix
-	Type  MatcherType `json:"type"`
-	Value string      `json:"value"`
+	Type  StatsMatchType `json:"type"`
+	Value string         `json:"value"`
 }
 
-type MatcherType string
+type StatsMatchType string
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_metric_types.go:71]: needs enum kubebuilder annotations

File: api/v1alpha1/envoyproxy_metric_types.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ type ProxyMetrics struct {
 	// `cluster.<cluster_name>.membership_degraded`，reference  https://github.com/envoyproxy/envoy/issues/9856,
 	// https://github.com/envoyproxy/envoy/issues/14610
 	//
-	Matches []Match `json:"matches,omitempty"`
+	Matches []StatsMatch `json:"matches,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/envoyproxy_metric_types.go:26]: replacing this with `sringsMatch`, should be backwards compatible from a YAML perspective, right ?

Comment (zhaohuabing) [reply api/v1alpha1/envoyproxy_metric_types.go]: Yes, we can. We should align the string match type across APIs. Currently supports exact, suffix, prefix, and Regex. Others can be included if needed in the future.

File: api/v1alpha1/envoyproxy_metric_types.go (hunk 2 above)
Comment (zhaohuabing) [reply api/v1alpha1/envoyproxy_metric_types.go]: Don't need it anymore :-)

**Review by zirain — APPROVED** [review]

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by arkodg** [comment]
should we merge this post v0.6 ?

**Comment by zhaohuabing** [comment]
> should we merge this post v0.6 ?
> 

I'm OK with that.

**Comment by zhaohuabing** [comment]
@arkodg can we merge it now? v0.6 has been released.
